### PR TITLE
GET-BOOKS: added implementation of DTO concept

### PR DESCRIPTION
**Pull request by kenu21**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -22,6 +23,11 @@
       <artifactId>spring-boot-starter</artifactId>
       <version>${spring-boot.version}</version>
```
Comment (Elena-Bruyako) [inline pom.xml]: remove everywhere, spring boot starter parent provides versions

```suggestion
```

File: src/main/java/com/keniu/config/package-info.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+/**
+ * This package contains the configuration
+ * classes for the book shop application.
+ * It includes global settings and configurations used across the application.
+ */
+package com.keniu.config;
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/config/package-info.java]: remove

File: src/main/java/com/keniu/controllers/BookController.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Controller class for managing book-related operations in the book shop application.
+ */
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/books")
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/controllers/BookController.java]:
```suggestion
@RequestMapping("/books")
```

File: src/main/java/com/keniu/controllers/package-info.java [hunk 4]
```diff
@@ -0,0 +1,4 @@
+/**
+ * This package contains the controllers for handling HTTP requests in the book shop application.
+ */
+package com.keniu.controllers;
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/controllers/package-info.java]: remove

File: src/main/java/com/keniu/dtos/package-info.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+/**
+ * This package contains Data Transfer Object (DTO)
+ * classes for the book shop application.
+ * DTOs are used to transfer data between
+ * the client and the server without exposing internal models.
+ */
+package com.keniu.dtos;
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/dtos/package-info.java]: remove

File: src/main/java/com/keniu/exceptions/EntityNotFoundException.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package com.keniu.exceptions;
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/exceptions/EntityNotFoundException.java:1]:
```suggestion
package com.keniu.exception;
```

File: src/main/java/com/keniu/dtos/CreateBookRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package com.keniu.dtos;
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/dtos/CreateBookRequestDto.java]:
```suggestion
package com.keniu.dto;
```

File: src/main/java/com/keniu/controllers/BookController.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
+package com.keniu.controllers;
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/controllers/BookController.java:1]:
```suggestion
package com.keniu.controller;
```

File: src/main/java/com/keniu/exceptions/package-info.java [hunk 9]
```diff
@@ -0,0 +1,4 @@
+/**
+ * This package contains the custom exceptions for the book shop application.
+ */
+package com.keniu.exceptions;
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/exceptions/package-info.java]: remove

File: src/main/java/com/keniu/mappers/package-info.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+/**
+ * This package contains mapper interfaces for the book shop application.
+ * Mappers are used to convert between
+ * entity models and Data Transfer Objects (DTOs),
+ * enabling separation between the application's data layer and its API layer.
+ */
+package com.keniu.mappers;
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/mappers/package-info.java]: remove

File: src/main/java/com/keniu/services/BookServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class BookServiceImpl implements BookService {
+    private final BookRepository bookRepository;
+    private final BookMapper bookMapper;
+
+    @Override
+    public BookDto save(CreateBookRequestDto createBookRequestDto) {
+        return bookMapper.toDto(bookRepository.save(bookMapper.toModel(createBookRequestDto)));
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/services/BookServiceImpl.java:20]: it's better to create local variable for Book `bookMapper.toModel(createBookRequestDto)`

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: pom.xml (hunk 1 above)
Comment (kenu21) [reply pom.xml]: Where you see "spring boot starter parent" in my pom?

File: src/main/java/com/keniu/controllers/BookController.java (hunk 8 above)
Comment (kenu21) [reply src/main/java/com/keniu/controllers/BookController.java:1]: Why?

File: src/main/java/com/keniu/services/BookServiceImpl.java (hunk 11 above)
Comment (kenu21) [reply src/main/java/com/keniu/services/BookServiceImpl.java:20]: For what?

File: src/main/java/com/keniu/config/package-info.java (hunk 2 above)
Comment (kenu21) [reply src/main/java/com/keniu/config/package-info.java]: Removed

File: src/main/java/com/keniu/controllers/BookController.java (hunk 3 above)
Comment (kenu21) [reply src/main/java/com/keniu/controllers/BookController.java]: Changed

File: src/main/java/com/keniu/controllers/package-info.java (hunk 4 above)
Comment (kenu21) [reply src/main/java/com/keniu/controllers/package-info.java]: Removed

File: src/main/java/com/keniu/dtos/package-info.java (hunk 5 above)
Comment (kenu21) [reply src/main/java/com/keniu/dtos/package-info.java]: Removed

File: src/main/java/com/keniu/exceptions/EntityNotFoundException.java (hunk 6 above)
Comment (kenu21) [reply src/main/java/com/keniu/exceptions/EntityNotFoundException.java:1]: Why?

File: src/main/java/com/keniu/dtos/CreateBookRequestDto.java (hunk 7 above)
Comment (kenu21) [reply src/main/java/com/keniu/dtos/CreateBookRequestDto.java]: Ranamed

File: src/main/java/com/keniu/exceptions/package-info.java (hunk 9 above)
Comment (kenu21) [reply src/main/java/com/keniu/exceptions/package-info.java]: Removed

File: src/main/java/com/keniu/mappers/package-info.java (hunk 10 above)
Comment (kenu21) [reply src/main/java/com/keniu/mappers/package-info.java]: Removed

File: pom.xml (hunk 1 above)
Comment (Elena-Bruyako) [reply pom.xml]: Oh, so the question why you don't have it? ) 

```
<parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
```

File: src/main/java/com/keniu/controllers/BookController.java (hunk 8 above)
Comment (Elena-Bruyako) [reply src/main/java/com/keniu/controllers/BookController.java:1]: Usually package name singular

File: src/main/java/com/keniu/repositories/BookRepositoryImpl.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            entityManager = entityManagerFactory.createEntityManager();
+            entityTransaction = entityManager.getTransaction();
+            entityManager.persist(book);
+            entityTransaction.commit();
+            return book;
+        } catch (Exception e) {
+            if (entityTransaction != null) {
+                entityTransaction.rollback();
+            }
+            throw new EntityNotFoundException("Can't save book " + book);
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/repositories/BookRepositoryImpl.java]: EntityNotFoundException is used in Service layer

```suggestion
            throw new DataProcessingException("Can't save book " + book);
```

File: src/main/java/com/keniu/repositories/BookRepositoryImpl.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            }
+        }
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return entityManager.createQuery("FROM book", Book.class).getResultList();
+        } catch (Exception e) {
+            throw new EntityNotFoundException("Can't get books");
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/repositories/BookRepositoryImpl.java]:
```suggestion
            throw new DataProcessingException("Can't get books");
```

File: src/main/java/com/keniu/repositories/BookRepositoryImpl.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return Optional.ofNullable(entityManager.find(Book.class, id));
+        } catch (Exception e) {
+            throw new EntityNotFoundException("can't get book with id " + id);
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/repositories/BookRepositoryImpl.java]:
```suggestion
            throw new DataProcessingException("Can't get book with id " + id);
```

File: src/main/java/com/keniu/services/BookServiceImpl.java (hunk 11 above)
Comment (Elena-Bruyako) [reply src/main/java/com/keniu/services/BookServiceImpl.java:20]: to improve readability and comfortable debugging
```suggestion
        Book book = bookMapper.toModel(createBookRequestDto);
        return bookMapper.toDto(bookRepository.save(book));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/keniu/controllers/BookController.java (hunk 8 above)
Comment (kenu21) [reply src/main/java/com/keniu/controllers/BookController.java:1]: Usually package name plural)

File: src/main/java/com/keniu/services/BookServiceImpl.java (hunk 11 above)
Comment (kenu21) [reply src/main/java/com/keniu/services/BookServiceImpl.java:20]: I don't set any value in it.
I only use it once.
The variable’s value carries no real meaning and doesn’t clarify anything to the reader.
In short, using the variable here is redundant.

File: pom.xml (hunk 1 above)
Comment (kenu21) [reply pom.xml]: So, your response is that you don’t see it in the POM, yet you suggest removing the versions?) OK, I've added parent.

File: src/main/java/com/keniu/repositories/BookRepositoryImpl.java (hunk 12 above)
Comment (kenu21) [reply src/main/java/com/keniu/repositories/BookRepositoryImpl.java]: Changes have been made

File: src/main/java/com/keniu/repositories/BookRepositoryImpl.java (hunk 13 above)
Comment (kenu21) [reply src/main/java/com/keniu/repositories/BookRepositoryImpl.java]: Changes have been made

File: src/main/java/com/keniu/repositories/BookRepositoryImpl.java (hunk 14 above)
Comment (kenu21) [reply src/main/java/com/keniu/repositories/BookRepositoryImpl.java]: Changes have been made

File: src/main/java/com/keniu/controllers/BookController.java (hunk 8 above)
Comment (Elena-Bruyako) [reply src/main/java/com/keniu/controllers/BookController.java:1]: don't agree with you. It's better to use singular abstract name. If you have a lot of controllers you can use subpackages to divide them

File: src/main/java/com/keniu/services/BookServiceImpl.java (hunk 11 above)
Comment (Elena-Bruyako) [reply src/main/java/com/keniu/services/BookServiceImpl.java:20]: it's only up to you, write readable code (not only for you but for other developers)

File: src/main/java/com/keniu/repositories/BookRepositoryImpl.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+            throw new DataProcessingException("Can't get books");
+        }
+    }
+
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return Optional.ofNullable(entityManager.find(Book.class, id));
+        } catch (Exception e) {
+            throw new DataProcessingException("can't get book with id " + id);
```
Comment (Elena-Bruyako) [inline src/main/java/com/keniu/repositories/BookRepositoryImpl.java]:
```suggestion
            throw new DataProcessingException("Can't get book with id " + id);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/main/java/com/keniu/controllers/BookController.java (hunk 8 above)
Comment (kenu21) [reply src/main/java/com/keniu/controllers/BookController.java:1]: I disagree with you. Controller refers to a single concept, whereas controllers implies multiple classes, which is better for scalability. In Java, it's common to name packages in the plural form when they contain groups of similar classes: services, models, utils. Of course, if the project already follows a different naming convention, it’s better to stick to the existing one. However, when starting a new project, it’s a matter of preference, but generally, using the plural form is the accepted convention.

File: src/main/java/com/keniu/services/BookServiceImpl.java (hunk 11 above)
Comment (kenu21) [reply src/main/java/com/keniu/services/BookServiceImpl.java:20]: Code is more readable when we avoid creating unnecessary variables.

File: src/main/java/com/keniu/repositories/BookRepositoryImpl.java (hunk 15 above)
Comment (kenu21) [reply src/main/java/com/keniu/repositories/BookRepositoryImpl.java]: Corrected

File: src/main/java/com/keniu/controllers/BookController.java (hunk 8 above)
Comment (okuzan) [reply src/main/java/com/keniu/controllers/BookController.java:1]: Matter of personal preference/project conventions. From my experience, the singular form is more common

File: src/main/java/com/keniu/services/BookServiceImpl.java (hunk 11 above)
Comment (okuzan) [reply src/main/java/com/keniu/services/BookServiceImpl.java:20]: Matter of personal preference/project conventions

**Review by okuzan — APPROVED** [review]

**Pull request merged**